File: apps/web/src/lib/phone.ts
/**
 * NANP-aware phone normalization, shared across every form that writes a
 * dialable phone number to the API. Moved out of
 * src/features/account/profileEdit.ts (campaign issue #234 PR 5, five
 * safety-review rounds — F2/F3/N1/R1/R3/R4 below) verbatim, unchanged
 * logic, so this exact reviewed behavior is reused rather than
 * reimplemented (issue #261: `backup_contact`'s phone gets the same
 * scrutiny as `landlords.phone` did there — it's the second number the
 * emergency chain dials, apps/api/app/agent/emergency_chain.py).
 * src/features/account/profileEdit.ts now imports from here instead of
 * defining its own copy.
 */

/**
 * CJK ideographic numerals (simple and financial/"trusted" forms across
 * Chinese/Japanese/Korean — e.g. 一二三四五六七八九十百千万億, plus their
 * CJK Compatibility Ideograph and Extension-B variants). Python's
 * `str.isnumeric()` (apps/api/app/phone.py's `_contains_non_ascii_digit`,
 * the server-side authority this file must not diverge from — #232/#260,
 * #273) treats these as numeric because Unicode gives them a
 * Numeric_Type=Numeric property, but their General_Category is `Lo`
 * (Letter, other), NOT one of the Number categories (`Nd`/`Nl`/`No`) — so
 * a JS `\p{N}` Unicode-property-escape test alone MISSES them; JS regex
 * has no Numeric_Type escape to fall back on. This literal list closes
 * that gap. It was generated (and diff-verified against Python's
 * `ch.isnumeric() and not ch.isascii()` across every codepoint
 * 0..0x10FFFF, zero misses) rather than hand-picked — see #273's PR
 * description for the generation script.
 */
const NON_ASCII_NUMERIC_RE =
  /[\p{Nd}\p{Nl}\p{No}\u{3405}\u{3483}\u{382A}\u{3B4D}\u{4E00}\u{4E03}\u{4E07}\u{4E09}\u{4E5D}\u{4E8C}\u{4E94}\u{4E96}\u{4EBF}\u{4EC0}\u{4EDF}\u{4EE8}\u{4F0D}\u{4F70}\u{5104}\u{5146}\u{5169}\u{516B}\u{516D}\u{5341}\u{5343}\u{5344}\u{5345}\u{534C}\u{53C1}\u{53C2}\u{53C3}\u{53C4}\u{56DB}\u{58F1}\u{58F9}\u{5E7A}\u{5EFE}\u{5EFF}\u{5F0C}\u{5F0D}\u{5F0E}\u{5F10}\u{62FE}\u{634C}\u{67D2}\u{6F06}\u{7396}\u{767E}\u{8086}\u{842C}\u{8CAE}\u{8CB3}\u{8D30}\u{9621}\u{9646}\u{964C}\u{9678}\u{96F6}\u{F96B}\u{F973}\u{F978}\u{F9B2}\u{F9D1}\u{F9D3}\u{F9FD}\u{20001}\u{20064}\u{200E2}\u{20121}\u{2092A}\u{20983}\u{2098C}\u{2099C}\u{20AEA}\u{20AFD}\u{20B19}\u{22390}\u{22998}\u{23B1B}\u{2626D}\u{2F890}]/u;

/**
 * `true` iff *value* contains a character that is numeric (by the test
 * above — a JS-reachable equivalent of Python's `str.isnumeric()`) but is
 * not plain ASCII `0`-`9`.
 *
 * Iterates by Unicode code point (`for...of` on a string), not UTF-16 code
 * unit, so a supplementary-plane character (e.g. the Extension-B ideograph
 * U+20001, above) is tested whole rather than as two broken surrogate
 * halves.
 */
function containsNonAsciiDigit(value: string): boolean {
  for (const ch of value) {
    if (ch.codePointAt(0)! > 0x7f && NON_ASCII_NUMERIC_RE.test(ch)) return true;
  }
  return false;
}

/**
 * Best-effort E.164 for the NANP inputs these forms actually see. Anything
 * this can't confidently normalize is rejected by `phoneLooksValid` below
 * rather than sent — storing a string Twilio can't dial is strictly worse
 * than refusing the edit, because the failure surfaces at 2am instead of
 * here.
 */
export function toE164(phone: string): string | null {
  const trimmed = phone.trim();
  // #273 (safety review, 2026-08-03 finding 1, client-side half — BLOCKING):
  // must run before ANY other processing, including the digit-strip below.
  // `.replace(/\D/g, "")` without the `u` flag is ASCII-only, so a
  // non-ASCII "digit" (Arabic-Indic "١", CJK "一"/"〇", fullwidth "４", …)
  // used to be treated as punctuation and silently STRIPPED rather than
  // rejected — shifting the remaining digits into a DIFFERENT, still-
  // plausible-looking number instead of failing loudly:
  // toE164("+١4165551234") used to return "+4165551234", a number that is
  // NOT the one the landlord typed, silently stored on `landlords.phone`
  // or `backup_contact.phone` — the field the emergency escalation chain
  // dials (apps/api/app/agent/emergency_chain.py). Reject outright
  // instead, mirroring apps/api/app/phone.py's `_contains_non_ascii_digit`
  // exactly (verified above) so the client never accepts what the server
  // would 422 on, and never silently rewrites what the server would
  // accept.
  if (containsNonAsciiDigit(trimmed)) return null;
  // R3 (safety re-verify): the +country test runs on the PUNCTUATION-
  // STRIPPED string, so "+44 20 7946 0958" is accepted the same as
  // "+442079460958" — a landlord with an international mobile shouldn't
  // have to guess our spacing rules.
  const plus = trimmed.startsWith("+");
  const digits = trimmed.replace(/\D/g, "");
  if (plus) {
    // N1 (safety re-verify): the international escape hatch must NOT
    // disable the NANP gate for NANP numbers. "+1416555013" — one digit
    // dropped, the likeliest typo there is — was being accepted and
    // stored, while the same typo without the plus was correctly
    // rejected. A +1 number gets exactly the same scrutiny as a bare one.
    if (digits.startsWith("1")) {
      return digits.length === 11 && isPlausibleNanp(digits.slice(1)) ? `+${digits}` : null;
    }
    return digits.length >= 8 && digits.length <= 15 ? `+${digits}` : null;
  }
  if (digits.length === 10 && isPlausibleNanp(digits)) return `+1${digits}`;
  if (digits.length === 11 && digits.startsWith("1") && isPlausibleNanp(digits.slice(1))) {
    return `+${digits}`;
  }
  // R1 (safety re-verify): returns NULL rather than `+${digits}`. The old
  // fallback could emit a bare "+" (or a digit-soup like "+416555013422"
  // from "416-555-0134 x22") onto the field emergency calls ring — it was
  // unreachable only because a guard in a THIRD file ran first, which is
  // exactly the validator/builder split that produced F2. Now no caller
  // can write an undialable value regardless of discipline.
  return null;
}

/** R4: NANP area code and exchange both start 2-9, and neither is an N11
 *  service code (411/911/…), which is never assignable (N3) — cheap
 *  regex that catches a dropped or mistyped digit before it becomes a
 *  permanently un-ringable emergency number. Deliberately permissive
 *  about everything else: rejecting a real number here is its own
 *  failure, since this is the field a landlord uses to be reachable. */
function isPlausibleNanp(digits: string): boolean {
  return /^(?!\d11)[2-9]\d{2}(?!\d11)[2-9]\d{6}$/.test(digits);
}

/**
 * Blank is valid ("keep my current number"). Anything NON-blank must carry
 * a dialable number.
 *
 * F2 (safety review, #234 PR 5): the old check counted digits only, so
 * zero-digit text — "n/a", "same as before", "-", exactly what a hurried
 * landlord types when the helper says "leave it blank to keep the number
 * already on file" — passed validation AND passed the builder's
 * `length > 0` send rule. The result was `landlords.phone = "n/a"`, a
 * green "Saved", and an emergency chain that silently never reaches
 * anyone. A non-empty field now has to look like a real number.
 */
export function phoneLooksValid(phone: string): boolean {
  const trimmed = phone.trim();
  if (trimmed.length === 0) return true;
  // Single source of truth with the normalizer (R1): valid means exactly
  // "toE164 can produce something dialable", so the two can never disagree
  // the way the pre-F2 pair did.
  return toE164(trimmed) !== null;
}

File: apps/api/app/phone.py
"""E.164 phone-number canonicalization — THE single source of truth for
what counts as a dialable phone number anywhere in this codebase (#232,
#260).

Every phone-bearing column in ``docs/03-engineering/schema-v1.md`` —
``landlords.phone``, ``properties.twilio_number``, the ``phone`` key inside
``properties.backup_contact``, ``tenants.phone``, ``vendors.phone`` — is
documented as E.164. Before this module existed, nothing enforced that at
write time: ``PATCH /v1/me`` accepted ``phone: ""`` (clearing the
emergency-call target as effectively as an explicit ``null``, #260) and no
endpoint validated shape at all, so a malformed value reached Twilio's
``create_call``/``send_sms`` and failed silently
(``app/agent/emergency_chain.py``'s ``_execute_action`` degrades a
bad-number failure to ``status='failed'`` by design — never re-raised), or
a format-drifted stored value simply never matched the Twilio ``/sms``
webhook's exact-string comparison, misrouting (or, since #170,
dead-lettering into ``unrouted_inbound``) a message that should have
reached a real property/tenant (#232).

``to_e164`` below is an ASCII-digits-only behavioral port of
``apps/web/src/features/account/profileEdit.ts``'s ``toE164`` (landed for
issue #234 PR 5) — client and server must never disagree about what is
dialable, for the ASCII input this codebase actually expects. Read that
file's own comments for the safety-review history (F2/F3/N1/R1/R3/R4)
that produced these exact rules; it is not repeated here.

**Non-ASCII "digit" characters are REJECTED, never silently stripped or
matched (safety review, 2026-08-03, finding 1 — BLOCKING).** Python's ``\\d``/
``\\D`` are Unicode-aware by default (``\\d`` matches the full Unicode
``Nd`` category — Arabic-Indic, Devanagari, fullwidth digits, …), unlike
JavaScript's, which are always ASCII-only. Left unguarded, a value like
``"+١4165551234"`` (one Arabic-Indic ``1`` ahead of an otherwise
ordinary Toronto number) would parse differently in each language: this
module's regexes now carry ``re.ASCII`` so ``\\d``/``\\D`` behave exactly
like JavaScript's (a non-ASCII digit is "not a digit" to either regex, so
it is stripped as punctuation would be, same as the TS side) — but
SILENTLY dropping what a human plausibly intended as a real digit is its
own hazard (it can shift the remaining digits into a DIFFERENT,
still-plausible-looking number instead of failing loudly). ``to_e164``
therefore rejects outright, before any other processing, if the input
contains ANY character that is a digit under Python's own (Unicode-aware)
``str.isdigit()`` but is not plain ASCII ``0``-``9`` — never silently
dropped, never coerced.

Canonicalization policy (mirrors the TS implementation exactly)
------------------------------------------------------------------------
- **Empty / whitespace-only** input -> rejected (``None``).
- **Bare NANP 10-digit** (e.g. ``"4165551234"``) -> assumed ``+1``; area
  code and exchange must both plausibly be NANP (see ``is_plausible_nanp``).
- **11-digit, leading ``1``** (e.g. ``"14165551234"``) -> ``+`` prepended,
  same NANP plausibility check on the trailing 10 digits.
- **Already ``+``-prefixed** -> a leading parenthesized trunk zero is
  dropped first (see "Parenthesized trunk zero" below), then punctuation
  is stripped, then:
  - if the remaining digits start with ``1``, treated as NANP: must be
    exactly 11 digits and pass the same plausibility check (a ``+1``
    number gets no less scrutiny than a bare one — a dropped digit like
    ``"+1416555013"`` must not slip through just because it has a plus).
  - otherwise treated as non-NANP international: accepted if the total
    digit count is between 8 and 15 inclusive (E.164's own bound), no
    further shape validation — this codebase has no non-NANP dialing
    rules to check against.
- **Extensions / any other junk** (e.g. ``"416-555-0134 x22"``, ``"n/a"``)
  -> rejected (``None``) — the digit-soup either fails the NANP
  plausibility check or the international length bound.

Parenthesized trunk zero (#277)
------------------------------------------------------------------------
``"+44 (0)20 7946 0958"`` — the single most common *written* form of a UK
number — used to normalize to ``"+4402079460958"``: a 13-digit non-number
Twilio rejects (21211), landing on the field the emergency chain dials
(``app/agent/emergency_chain.py``'s ``_execute_action`` swallows that
failure into ``status="failed"`` without re-raising, so the failure is
silent). The ``(0)`` is a trunk prefix, dialed only for calls placed
*inside* the country and never part of the number in E.164 form.

This module now drops a leading trunk ``0`` when it appears
**parenthesized, directly after the country code** — narrow and
well-understood, not a general trunk-prefix rule for every country. It
must run on the RAW string, before punctuation is stripped: once digits
are collapsed a real ``(0)`` is indistinguishable from an ordinary digit
``0`` that happens to sit there. The match is anchored to the very start
of the ``+``-prefixed string (immediately after the country code) and
requires the literal parenthesized ``(0)`` (spaces/hyphens allowed
around it, per the anchor) — ``"+1 (416) 555 0100"`` is untouched: those
parentheses hold an area code, not a trunk marker, so the literal
``"(0)"`` never matches ``"(416)"``. A genuine, un-parenthesized leading
``0`` (e.g. ``"+44 020 7946 0958"``) is left exactly as before — this is
option 1 only (see the issue); it does not attempt to detect or repair
that case.

``is_plausible_nanp``: area code and exchange code both start ``2``-``9``
and neither is an N11 service code (``211``/``411``/``911``/… — never
assignable to a real subscriber) — the same cheap regex as the TS version,
deliberately permissive about everything else (a false REJECT here is its
own failure mode, since this is the field a landlord/tenant/vendor is
reached on).
"""

from __future__ import annotations

import re

from app.errors import AppError

# re.ASCII (safety-review finding, #232/#260 follow-up): without it, \d/\D
# are Unicode-aware (\d matches the full Nd category — Arabic-Indic,
# Devanagari, fullwidth digits, …), unlike JavaScript's, which are always
# ASCII-only — see module docstring, "Non-ASCII 'digit' characters".
_NON_DIGIT_RE = re.compile(r"\D+", re.ASCII)
_NANP_RE = re.compile(r"^(?!\d11)[2-9]\d{2}(?!\d11)[2-9]\d{6}$", re.ASCII)
# #277: anchored to the start of the string — "the country code" means
# whatever digit run sits between "+" and the parenthesized "(0)", nothing
# more. 1-3 digits mirrors E.164's own country-code length bound. Requires
# the LITERAL "(0)" (not just any parenthesized digit), which is what keeps
# "+1 (416) 555 0100" (an area code, not a trunk marker) untouched.
_PARENTHESIZED_TRUNK_ZERO_RE = re.compile(r"^(\+\d{1,3})[\s-]*\(0\)[\s-]*", re.ASCII)


def is_plausible_nanp(digits: str) -> bool:
    """``True`` iff *digits* (exactly 10 digits, no country code) is a
    plausible NANP subscriber number — see module docstring."""
    return _NANP_RE.match(digits) is not None


def _contains_non_ascii_digit(value: str) -> bool:
    """``True`` iff *value* contains a character Python considers numeric
    but which is not plain ASCII ``0``-``9`` — e.g. ``"١"`` (Arabic-Indic
    one), ``"٤"`` (Arabic-Indic four), ``"４"`` (fullwidth four).

    Uses ``str.isnumeric()`` rather than ``str.isdigit()`` (safety review,
    2026-08-03, finding 1 residual): ``isdigit()`` misses CJK numerals —
    ``"一"`` and ``"〇"`` are literally one and zero, i.e. exactly "a
    character a human plausibly meant as a digit", which is this guard's
    whole rationale. Without this they were silently STRIPPED and the
    remainder accepted, so ``"+一4165551234"`` became ``"+4165551234"`` —
    a different, still-plausible number stored on the field the escalation
    chain dials. The ``not ch.isascii()`` conjunct keeps ``0``-``9`` out of
    it, and no legitimate phone character (``+ - ( ) . space x``) is
    numeric, so this rejects nothing real."""
    return any(ch.isnumeric() and not ch.isascii() for ch in value)


def _drop_parenthesized_trunk_zero(value: str) -> str:
    """Drop a leading trunk ``0`` written parenthesized directly after the
    country code (``"+44 (0)20 7946 0958"`` -> ``"+4420 7946 0958"``) —
    see module docstring, "Parenthesized trunk zero" (#277). Must be
    called on the RAW (punctuation-intact) string, before any digit
    stripping — once digits are collapsed a real ``(0)`` can no longer be
    told apart from an ordinary digit ``0``. A no-op when the pattern
    doesn't match (e.g. ``"+1 (416) 555 0100"``, where the parenthesized
    content is an area code, not a literal ``"0"``)."""
    return _PARENTHESIZED_TRUNK_ZERO_RE.sub(r"\1", value, count=1)


def to_e164(raw: str) -> str | None:
    """Best-effort canonicalization of *raw* to E.164, or ``None`` if it
    cannot be confidently canonicalized — see module docstring for the
    exact policy.

    Pure, no I/O — safe to call from a route handler, a migration, or the
    Twilio webhook's routing-match path alike.
    """
    if _contains_non_ascii_digit(raw):
        # Never silently strip/ignore a non-ASCII digit character —
        # rejecting outright is the only safe response (see module
        # docstring); checked before anything else, including .strip().
        return None

    trimmed = raw.strip()
    plus = trimmed.startswith("+")
    if plus:
        # #277: only the international branch — must run before the
        # punctuation strip below, and only on a "+"-prefixed value (a
        # bare NANP input has no country code for "(0)" to sit after).
        trimmed = _drop_parenthesized_trunk_zero(trimmed)
    digits = _NON_DIGIT_RE.sub("", trimmed)

    if plus:
        if digits.startswith("1"):
            return f"+{digits}" if len(digits) == 11 and is_plausible_nanp(digits[1:]) else None
        return f"+{digits}" if 8 <= len(digits) <= 15 else None

    if len(digits) == 10 and is_plausible_nanp(digits):
        return f"+1{digits}"
    if len(digits) == 11 and digits.startswith("1") and is_plausible_nanp(digits[1:]):
        return f"+{digits}"
    return None


def canonicalize_phone(value: str, *, field: str) -> str:
    """Canonicalize *value* to E.164 or raise 422 ``invalid_field``.

    ``value`` here is already known to be non-null — callers that need to
    treat an explicit empty string as equivalent to ``null`` for a
    not-nullable-by-business-rule field must do so BEFORE calling this
    (see ``app.validation.normalize_blank_to_null``, then
    ``app.validation.reject_explicit_null``). An empty string reaching
    this function directly (e.g. a required-field create body) is simply
    uncanonicalizable, same as any other junk input, and 422s the same
    way — *field* is a fixed, known field name (e.g. ``"phone"`` or
    ``"backup_contact.phone"``), never request data, so this stays within
    ``AppError``'s static-message rule (``app/errors.py``).
    """
    canonical = to_e164(value)
    if canonical is None:
        raise AppError(
            status_code=422,
            code="invalid_field",
            message=f"{field} must be a valid, dialable phone number.",
        )
    return canonical


__all__: list[str] = ["canonicalize_phone", "is_plausible_nanp", "to_e164"]
